django/test/testcases.py

```diff
@@ -488,142 +589,94 @@ def _check_test_client_response(self, response, attribute, method_name):
-                        self.assertTrue(
-                            err in field_errors,
-                            msg_prefix + "The field '%s' on form '%s' in"
-                            " context %d does not contain the error '%s'"
-                            " (actual errors: %s)" %
-                            (field, form, i, err, repr(field_errors))
-                        )
-                    elif field in context[form].fields:
-                        self.fail(
-                            msg_prefix + "The field '%s' on form '%s' in context %d contains no errors" %
-                            (field, form, i)
-                        )
-                    else:
-                        self.fail(
-                            msg_prefix + "The form '%s' in context %d does not contain the field '%s'" %
-                            (form, i, field)
-                        )
-                else:
-                    non_field_errors = context[form].non_field_errors()
-                    self.assertTrue(
-                        err in non_field_errors,
-                        msg_prefix + "The form '%s' in context %d does not"
-                        " contain the non-field error '%s'"
-                        " (actual errors: %s)" %
-                        (form, i, err, non_field_errors or 'none')
-                    )
-        if not found_form:
-            self.fail(msg_prefix + "The form '%s' was not used to render the response" % form)
+        self.assertEqual(field_errors, errors, msg_prefix + failure_message)
 
-    def assertFormsetError(self, response, formset, form_index, field, errors,
-                           msg_prefix=''):
+    @_assert_form_error_deprecation_helper.patch_signature  # RemovedInDjango50Warning (the decorator)
+    def assertFormError(self, form, field, errors, msg_prefix=''):
         """
-        Assert that a formset used to render the response has a specific error.
+        Assert that a field named `field` on the given form object has specific errors.
 
-        For field errors, specify the ``form_index`` and the ``field``.
-        For non-field errors, specify the ``form_index`` and the ``field`` as
-        None.
-        For non-form errors, specify ``form_index`` as None and the ``field``
-        as None.
+        errors can be either a single error (string) or a list of errors.
+        Using errors=[] will test that the field has no errors.
+
+        You can pass field=None to check the form's non-field errors.
         """
-        self._check_test_client_response(response, 'context', 'assertFormsetError')
-        # Add punctuation to msg_prefix
+        if errors is None:
+            warnings.warn(
+                "Passing errors=None to assertFormError() is deprecated. Use errors=[] instead.",
+                RemovedInDjango50Warning,
+                stacklevel=2
+            )
+            errors = []
```

Comment:
Can we deprecate passing `None` in `errors` in a separate PR? and do this in advance. There is already a lot of changes in this patch, that should make it more readable and easier to review. I can try to refactor this out if you don't have time to keep working on this.

Reply:
That's fair, I can rework the patch and open a separate ticket for deprecating `errors=None`.

Reply:
Thanks :+1: , IMO a separate ticket is not necessary, we can `Refs #33348 -- ...`.

Reply:
I've opened a new PR with only the `errors=None` deprecation: #15282

I will put this current PR on hold until this new one is merged.